surya/common/surya/__init__.py

```diff
@@ -580,4 +591,173 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
+            (valid_batch_size, max_grid_size, pixel_values.shape[-1]),
+            dtype=pixel_values.dtype,
+        )
+
+        # Roll out into a full grid
+        seq_len = 0
+        row_grids = []
+        for i in range(valid_batch_size):
+            curr_sample_len = grid_thw[i][0] * grid_thw[i][1] * grid_thw[i][2]
+            full_image_grid[i, -curr_sample_len:] = pixel_values[
+                seq_len : seq_len + curr_sample_len
+            ]
+            padded_len = max_grid_size - curr_sample_len
+            if padded_len > 0:
+                row_grid = torch.tensor(
+                    [
+                        [1, 4, padded_len // 4],
+                        grid_thw[i].tolist(),
+                    ],
+                    dtype=torch.long,
+                )
+            else:
+                row_grid = torch.tensor(
+                    [
+                        grid_thw[i].tolist(),
+                    ],
+                    dtype=torch.long,
+                )
+
+            row_grids.append(row_grid)
+            seq_len += curr_sample_len
+
+        # bsz, 2, 3
+        row_grids = torch.stack(row_grids, dim=0)
+
+        if settings.FOUNDATION_STATIC_CACHE:
+            # Pad to max batch size, repeat the final row
+            row_grids = pad_to_batch_size_repeat(
+                row_grids,
+                batch_size=max_batch_size,
+            )
+            full_image_grid = pad_to_batch_size(
+                full_image_grid,
+                batch_size=max_batch_size,
+            )
+
+        full_image_grid = full_image_grid.to(self.device)
+
+        embeddings = self.vision_encoder.embed_images(
+            image_batch=full_image_grid, grid_thw=row_grids.to(self.device)
+        )
+
+        encoding_2d = self.get_2d_learned_embeddings(
+            row_grids,
+            bbox_size=self.config.image_embed_encoding_multiplier,
+        )
+        embeddings += encoding_2d
+        return embeddings
+
+    def embed_ids_boxes_images(
+        self,
+        input_ids,
+        image_embeddings,
+        encoder_chunk_size: int,
+        valid_batch_size: torch.Tensor | None = None,
+    ):
+        """
+        Insert embedded image tiles into the corresponding positions into the full input sequence
+
+        Positions to insert new tokens are indicated by the special image token index
+        """
+        # This is batched in the inner call
+        inputs_embeds = self.embedder.embed(input_tokens=input_ids)
+
+        if image_embeddings is not None:
+            image_token_id_tensor = torch.tensor(
+                self.config.image_token_id,
+                device=inputs_embeds.device,
+                dtype=torch.long,
+            )
+            mask = input_ids == image_token_id_tensor
+            last_image_token_pos = (
+                mask.size(1)
+                - 1
+                - mask.flip(dims=[1]).long().argmax(dim=1, keepdim=True)
+            )
```

Comment:
Slightly confusing - Can't we just do `torch.argmax(dim=1, keepdim=True)`? I might be missing something though

Reply:
argmax gives the first image token, but we need the last